### DISCUSSION TITLE
Improve tests

### DISCUSSION THREAD

**Discussion by wwerner** [gr3p1p3/transparent-proxy#29] [Ideas]
Hey again, 

I'd love to swap out the hand-rolled unit test solution for a framework.
Drivers:
* Tests use external systems (e.g. ifconfig.me, httpbin.org), so they depend on those services to be up and available
* It's hard to track coverage
* Adding tests is a bit cumbersome, at some point the test file will grow large, then we'll need to regorganize tests anyway.
* Tests rely on `curl` to be installed locally. This might not be the case by default when building in containers. I also had [issues w/ a different `curl` version](https://github.com/gr3p1p3/transparent-proxy/commit/22b9e76f371f3b04a814bd23ed8932f50824d594).

I've tried with [Jest](https://jestjs.io/) in my (very WIP) TS port. 
Here's the gist of it:
* Instead of targeting external systems, the tests spin up a tiny local express app that basically returns the received HTTP messages. See [here](https://github.com/wwerner/transparent-proxy/pull/2/files#diff-d28a461c46dd3d6440988800fbf174825ddbec8b60488095df612d33f1cad4df).
* Use a utility function to make requests against this app through the proxy using plain node `http`: https://github.com/wwerner/transparent-proxy/pull/2/files#diff-8951d3f508f54ec05c5df49f04f821a2aec5c75bce858639461f388e498a398cR10
* Tests then spin up a configured proxy, make a request and assert the results: https://github.com/wwerner/transparent-proxy/pull/2/files#diff-05d1e5650843d2a077ab7636ce95d7af53569eb7322319940c67fd5d89fe71a9R4
* Jest allows for [measuring coverage](https://jestjs.io/docs/cli#--coverageboolean), is well supported, provides [mocking primitives](https://jestjs.io/docs/mock-functions) and nicely extensible with custom [matchers](https://jestjs.io/docs/mock-functions#custom-matchers), should we want to.

I could help switch the existing tests over.
What do you think?

**Comment by gr3p1p3** [comment]
Yess! I thought multiple times about the use of curl and refactoring the test.js file that is growing slowly up.
I'm not confidential with jest, but I can learn :) 

Thanks for this! Let's do it!

**Reply by wwerner** [reply]
I pushed a scaffold for testing and one initial test here: https://github.com/gr3p1p3/transparent-proxy/pull/31 

This is what a test will look like in the new structure: https://github.com/gr3p1p3/transparent-proxy/pull/31/files#diff-3a306758c1d18927d16b133c9fc27dae8661649521e55023f03ce846974bc275R4

I'll continue working on porting the rest of the tests.

@leppert: I'm still unsure whether I want to maintain a fork. Until we've decided, I'll continue contributing here.
